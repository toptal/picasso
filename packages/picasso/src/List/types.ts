export type ListVariant = 'ordered' | 'unordered'
