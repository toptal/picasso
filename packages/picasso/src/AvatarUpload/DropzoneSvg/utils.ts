/**
 * Returns the shape of the Avatar component in SVG path's format.
 * SVG's "path" works for the function like this:
 * M x1 y1 - move to point x, y (start point)
 * H x2 - horizontal line to x1 (top edge)
 * V y3 - vertical line to y2 (right edge)
 * H x4 - horizontal line to x3 (bottom edge)
 * L x1 y5 - close path to point x4, y4
 * Z - draw path back to starting point (x1 y1)
 */
const getAvatarShape = ({
  bottomEdgeLength,
  startPoint,
  leftCornerPoint,
  rightEdgeLength,
  topEdgeLength,
  drawBackToStart,
}: {
  startPoint: number
  topEdgeLength: number
  rightEdgeLength: number
  bottomEdgeLength: number
  leftCornerPoint: number
  drawBackToStart: boolean
}) => {
  return `
    M ${startPoint} ${startPoint}
    H ${topEdgeLength} 
    V ${rightEdgeLength} 
    H ${bottomEdgeLength} 
    L ${startPoint} ${leftCornerPoint}${drawBackToStart ? ' Z' : ''}
  `
}

/**
 * Returns the shape of the background of the avatar.
 * For small variant, it should be 80x80(px), for large - 160x160(px).
 */
export const getBackgroundShape = ({
  dimensions,
  cornerSize,
}: {
  dimensions: number
  cornerSize: number
}) => {
  const centerShift = 3 // shift for the outline stroke

  return getAvatarShape({
    startPoint: centerShift,
    topEdgeLength: dimensions + centerShift,
    rightEdgeLength: dimensions + centerShift,
    bottomEdgeLength: cornerSize + centerShift,
    leftCornerPoint: dimensions - cornerSize + centerShift,
    drawBackToStart: false,
  })
}

/**
 * Returns the shape of the outline when field is focused.
 * For small variant, it should be 82x82(px), for large - 162x162(px).
 * it is 2px bigger than the background because of the outline stroke width.
 */
export const getOutlineShape = ({
  dimensions,
  cornerSize,
}: {
  dimensions: number
  cornerSize: number
}) => {
  const centerShift = 2 // shift for the outline stroke
  const outlineStrokeWidth = 2 // width of the outline stroke

  return getAvatarShape({
    startPoint: centerShift,
    topEdgeLength: dimensions + centerShift + outlineStrokeWidth,
    rightEdgeLength: dimensions + centerShift + outlineStrokeWidth,
    bottomEdgeLength: cornerSize + 1,
    leftCornerPoint: dimensions - cornerSize + centerShift,
    drawBackToStart: true,
  })
}

/**
 * Returns the shape of the borders.
 * For small variant, it should be 78x78(px), for large - 158x158(px).
 * it is 1px smaller than the background because of the border stroke width.
 */
export const getBordersShape = ({
  dimensions,
  cornerSize,
}: {
  dimensions: number
  cornerSize: number
}) => {
  const centerShift = 4 // shift for the outline stroke and border stroke
  const outlineStrokeWidth = 2 // width of the outline stroke

  return getAvatarShape({
    startPoint: centerShift,
    topEdgeLength: dimensions + outlineStrokeWidth,
    rightEdgeLength: dimensions + outlineStrokeWidth,
    bottomEdgeLength: cornerSize + outlineStrokeWidth,
    leftCornerPoint: dimensions - cornerSize + outlineStrokeWidth,
    drawBackToStart: true,
  })
}
