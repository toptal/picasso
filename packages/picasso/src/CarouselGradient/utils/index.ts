export { default } from './gradient-width'
