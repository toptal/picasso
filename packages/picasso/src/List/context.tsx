import React, { createContext, useContext } from 'react'

export type ListItemType =
  | 'circle'
  | 'disc'
  | 'checkmark'
  | 'arrow'
  | 'numeral'
  | 'alpha'
  | 'roman'

export interface ListContextType {
  styleType?: ListItemType
  /** Level of list nesting, when not a child of a list it has level 0 */
  level: number
}

const ListContext = createContext<ListContextType>({
  level: 0,
})

export const useListContext = () => useContext(ListContext)

export interface ListContextProviderProps {
  styleType?: ListItemType
  children: React.ReactNode
}

export const ListContextProvider = ({
  children,
  styleType,
}: ListContextProviderProps) => {
  const { level } = useListContext()

  return (
    <ListContext.Provider value={{ level: level + 1, styleType }}>
      {children}
    </ListContext.Provider>
  )
}
