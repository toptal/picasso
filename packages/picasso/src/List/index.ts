import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './List'

export { default } from './List'
export type ListProps = OmitInternalProps<Props>

export { useListContext } from './context'
export type { ListContextType, ListItemType } from './context'
