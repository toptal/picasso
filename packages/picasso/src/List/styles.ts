import { createStyles, Theme } from '@material-ui/core/styles'

export default ({ palette }: Theme) =>
  createStyles({
    root: {
      listStyle: 'none',
      padding: 0,
      margin: 0,
      fontSize: '1rem',
    },
    unordered: {
      color: palette.text.primary,
    },
  })
