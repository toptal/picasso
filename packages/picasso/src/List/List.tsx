import React, { ReactNode } from 'react'
import { BaseProps } from '@toptal/picasso-shared'
import { Theme, makeStyles } from '@material-ui/core/styles'
import cx from 'classnames'

import ListItem from '../ListItem'
import styles from './styles'

export type Props = BaseProps & {
  children: ReactNode
  /** The variant to use */
  variant: 'ordered' | 'unordered'
  /** Specifies the start value of the first list item in an ordered list */
  start?: number
}

const useStyles = makeStyles<Theme>(styles, { name: 'PicassoList' })

const Tags = {
  unordered: 'ul',
  ordered: 'ol'
} as const

export const List = (props: Props) => {
  const classes = useStyles()
  const { variant, children, start = 1, className, ...rest } = props

  const totalChildElements = React.Children.count(children)

  const listItems = React.Children.map(children, (child, index) => {
    if (!React.isValidElement(child)) {
      return child
    }

    return React.cloneElement(child, {
      variant,
      isLastElement: totalChildElements === index + 1,
      index: index + start - 1
    })
  })

  const ListTag = Tags[variant]

  return (
    <ListTag
      className={cx(classes.root, classes[variant], className)}
      {...rest}
    >
      {listItems}
    </ListTag>
  )
}

List.defaultProps = {
  variant: 'unordered',
  start: 1
}

List.Item = ListItem

List.displayName = 'List'

export default List
