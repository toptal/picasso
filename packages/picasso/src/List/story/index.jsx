import { List } from '../List'
import listItemStory from '../../ListItem/story'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage('List', null)

page
  .createTabChapter('Props')
  .addComponentDocs({
    component: List,
    name: 'List',
    additionalDocs: {
      variant: {
        name: 'variant',
        defaultValue: 'unordered',
        type: {
          name: 'enum',
          enums: ['ordered', 'unordered'],
        },
      },
      start: {
        name: 'start',
        defaultValue: '1',
        type: {
          name: 'number',
        },
        description:
          'Specifies the start value of the first list item in an ordered list',
      },
    },
  })
  .addComponentDocs(listItemStory.componentDocs)

page
  .createChapter()
  .addExample('List/story/Ordered.example.tsx', {
    title: 'Ordered',
    takeScreenshot: false,
  })
  .addExample('List/story/OrderedWithStart.example.tsx', {
    title: 'Ordered with custom start',
    takeScreenshot: false,
  })
  .addExample('List/story/Unordered.example.tsx', {
    title: 'Unordered',
    takeScreenshot: false,
  })
  .addExample('List/story/Custom.example.tsx', {
    title: 'Custom',
    takeScreenshot: false,
  })
