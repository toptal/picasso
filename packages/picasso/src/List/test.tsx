import React from 'react'
import { render } from '@toptal/picasso/test-utils'
import { OmitInternalProps } from '@toptal/picasso-shared'

import List, { Props } from './List'
import { generateListItems } from './utils'

const testIds = {
  ul: 'unordered-list',
  ol: 'ordered-list',
}

const renderList = (props: OmitInternalProps<Props>) =>
  render(<List {...props}>{props.children}</List>)

describe('List', () => {
  it('renders unordered list', () => {
    const { container, getByTestId } = renderList({
      children: generateListItems(5),
      variant: 'unordered',
      'data-testid': testIds.ul,
    })

    const list = getByTestId(testIds.ul) as HTMLUListElement

    expect(list.tagName).toBe('UL')
    expect(list.children).toHaveLength(5)
    expect(container).toMatchSnapshot()
  })

  it('renders ordered list', () => {
    const { container, getByTestId } = renderList({
      children: generateListItems(5),
      variant: 'ordered',
      'data-testid': testIds.ol,
    })

    const list = getByTestId(testIds.ol) as HTMLOListElement

    expect(list.tagName).toBe('OL')
    expect(list.children).toHaveLength(5)
    expect(container).toMatchSnapshot()
  })
})
