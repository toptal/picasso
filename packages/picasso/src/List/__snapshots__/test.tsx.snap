// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders ordered list 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <ol
      class="PicassoList-root"
      data-testid="ordered-list"
    >
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-ordered"
          >
            <p
              class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
            >
              1
              .
            </p>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N1
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-ordered"
          >
            <p
              class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
            >
              2
              .
            </p>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N2
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-ordered"
          >
            <p
              class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
            >
              3
              .
            </p>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N3
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-ordered"
          >
            <p
              class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
            >
              4
              .
            </p>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N4
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer PicassoListItem-lastElement"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-ordered"
          >
            <p
              class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
            >
              5
              .
            </p>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N5
          </div>
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders unordered list 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <ul
      class="PicassoList-root PicassoList-unordered"
      data-testid="unordered-list"
    >
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-unordered"
          >
            <svg
              class="PicassoSvgBullet16-root"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="M8 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4Z"
              />
            </svg>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N1
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-unordered"
          >
            <svg
              class="PicassoSvgBullet16-root"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="M8 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4Z"
              />
            </svg>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N2
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-unordered"
          >
            <svg
              class="PicassoSvgBullet16-root"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="M8 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4Z"
              />
            </svg>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N3
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-unordered"
          >
            <svg
              class="PicassoSvgBullet16-root"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="M8 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4Z"
              />
            </svg>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N4
          </div>
        </div>
      </li>
      <li>
        <div
          class="PicassoContainer-flex PicassoListItem-listContainer PicassoListItem-lastElement"
        >
          <div
            class="PicassoContainer-rightsmallMargin PicassoContainer-flexEndJustifyContent PicassoContainer-inline PicassoListItem-unordered"
          >
            <svg
              class="PicassoSvgBullet16-root"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="M8 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4Z"
              />
            </svg>
          </div>
          <div
            class="MuiTypography-root PicassoTypography-bodyMedium MuiTypography-body1"
          >
            list item N5
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
